feat(mcp): tool request callback (#2000)
pydantic/pydantic-ai: docs/mcp/client.md

1. See [MCP Run Python](run-python.md) for more information.

## Tool call customisation

The MCP servers provide the ability to set a `process_tool_call` which allows
the customisation of tool call requests and their responses.

A common use cas for this is to inject metadata to the requests which the server
call needs.

> ```python {title="mcp_process_tool_call.py" py="3.10"}

Reviewer: Why 3.10 specifically?

Author: This was copied from the other examples

from typing import Any

from pydantic_ai import Agent
from pydantic_ai.mcp import CallToolFunc, MCPServerStdio, ToolResult
from pydantic_ai.models.test import TestModel
from pydantic_ai.tools import RunContext

async def process_tool_call(
    ctx: RunContext[int],

>     tool_call: CallToolFunc,

Reviewer: Let's call it `call_tool`

```suggestion
    call_tool: CallToolFunc,
```

    tool_name: str,
    args: dict[str, Any],
) -> ToolResult:

>     """A noop process_tool_call that just sets a flag."""

Reviewer: Not a noop since it passes the entire deps to the server!

Author: Evolved, will update.

Reviewer: ```suggestion
    """A tool call processor that passes along the deps."""
```

>     return await tool_call(tool_name, args, metadata={'run_context': ctx.deps})

Reviewer: It's not the full run context, so let's call this `deps`

```suggestion
    return await tool_call(tool_name, args, metadata={'deps': ctx.deps})
```
